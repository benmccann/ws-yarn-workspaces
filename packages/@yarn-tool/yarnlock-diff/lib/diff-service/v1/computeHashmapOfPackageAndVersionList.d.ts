import { IPackageData } from './types';
import { IComputedPackage } from '../types';
export declare const PACKAGE_REGEX: RegExp;
export declare function computeHashmapOfPackageAndVersionList(alreadyComputedPackage: IComputedPackage, parsedOldPackage: Record<string, IPackageData>): IComputedPackage;
