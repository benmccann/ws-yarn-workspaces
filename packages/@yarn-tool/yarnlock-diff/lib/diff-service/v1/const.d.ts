export declare const PACKAGE_REGEX: RegExp;
