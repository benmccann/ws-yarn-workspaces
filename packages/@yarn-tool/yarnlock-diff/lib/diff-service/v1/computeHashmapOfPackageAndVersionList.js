"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.computeHashmapOfPackageAndVersionList = exports.PACKAGE_REGEX = void 0;
exports.PACKAGE_REGEX = /(?<packageName>.*)@(?:(?<semverPin>[\^\$])?(?<major>\d)(?:\.(?<minor>\d))?(?:\.(?<patch>\d))?(?:-(?<prerelease>[0-9a-zA-Z-]+)(?:.(?<prereleaseVersion>[0-9a-zA-Z]+))?)?(?:\+(?<metadata>[0-9a-zA-Z-]+)(?:.(?<metadataVersion>[0-9a-zA-Z]+))?)?|\*)/;
function computeHashmapOfPackageAndVersionList(alreadyComputedPackage, parsedOldPackage) {
    const newComputedPackage = { ...alreadyComputedPackage };
    Object.entries(parsedOldPackage).forEach(([packageName, packageData]) => {
        const regexResult = exports.PACKAGE_REGEX.exec(packageName);
        const packageNameWithoutVersion = regexResult && regexResult.groups && regexResult.groups.packageName;
        if (!packageNameWithoutVersion)
            return;
        if (newComputedPackage[packageNameWithoutVersion]) {
            newComputedPackage[packageNameWithoutVersion].push(packageData.version);
            newComputedPackage[packageNameWithoutVersion] = [
                ...new Set(newComputedPackage[packageNameWithoutVersion]),
            ];
            newComputedPackage[packageNameWithoutVersion].sort();
        }
        else {
            newComputedPackage[packageNameWithoutVersion] = [packageData.version];
        }
    });
    return newComputedPackage;
}
exports.computeHashmapOfPackageAndVersionList = computeHashmapOfPackageAndVersionList;
//# sourceMappingURL=computeHashmapOfPackageAndVersionList.js.map