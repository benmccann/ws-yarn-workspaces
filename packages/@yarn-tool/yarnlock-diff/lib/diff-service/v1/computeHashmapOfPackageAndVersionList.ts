import { IPackageData } from './types';
import { IComputedPackage } from '../types';

export const PACKAGE_REGEX = /(?<packageName>.*)@(?:(?<semverPin>[\^\$])?(?<major>\d)(?:\.(?<minor>\d))?(?:\.(?<patch>\d))?(?:-(?<prerelease>[0-9a-zA-Z-]+)(?:.(?<prereleaseVersion>[0-9a-zA-Z]+))?)?(?:\+(?<metadata>[0-9a-zA-Z-]+)(?:.(?<metadataVersion>[0-9a-zA-Z]+))?)?|\*)/;

export function computeHashmapOfPackageAndVersionList(
	alreadyComputedPackage: IComputedPackage,
	parsedOldPackage: Record<string, IPackageData>,
): IComputedPackage
{
	const newComputedPackage = { ...alreadyComputedPackage };
	Object.entries(parsedOldPackage).forEach(([packageName, packageData]) =>
	{
		const regexResult = PACKAGE_REGEX.exec(packageName);
		const packageNameWithoutVersion =
			regexResult && regexResult.groups && regexResult.groups.packageName;

		if (!packageNameWithoutVersion) return;
		if (newComputedPackage[packageNameWithoutVersion])
		{
			newComputedPackage[packageNameWithoutVersion].push(packageData.version);
			newComputedPackage[packageNameWithoutVersion] = [
				...new Set(newComputedPackage[packageNameWithoutVersion]),
			];
			newComputedPackage[packageNameWithoutVersion].sort();
		}
		else
		{
			newComputedPackage[packageNameWithoutVersion] = [packageData.version];
		}
	});
	return newComputedPackage;
}
