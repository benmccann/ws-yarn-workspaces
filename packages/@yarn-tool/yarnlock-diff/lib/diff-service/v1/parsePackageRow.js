"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parsePackageRow = void 0;
const const_1 = require("./const");
function parsePackageRow(packageName, packageData) {
    var _a;
    const regexResult = const_1.PACKAGE_REGEX.exec(packageName);
    const packageNameWithoutVersion = (_a = regexResult === null || regexResult === void 0 ? void 0 : regexResult.groups) === null || _a === void 0 ? void 0 : _a.packageName;
    if (packageNameWithoutVersion) {
        return {
            name: packageNameWithoutVersion,
            version: packageData.version,
        };
    }
}
exports.parsePackageRow = parsePackageRow;
//# sourceMappingURL=parsePackageRow.js.map