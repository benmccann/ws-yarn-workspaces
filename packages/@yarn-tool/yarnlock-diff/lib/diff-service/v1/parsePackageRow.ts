import { IPackageData } from './types';
import { PACKAGE_REGEX } from './const';
import { IParsePackageRow } from '../types';
import npa, { Result } from 'npm-package-arg';

export function parsePackageRow(packageName: string, packageData: IPackageData): IParsePackageRow
{
	let parsed: Result;

	try
	{
		parsed = npa(packageName);
	}
	catch (e) {}

	const packageNameWithoutVersion =
		parsed?.name ?? PACKAGE_REGEX.exec(packageName)?.groups?.packageName;

	if (packageNameWithoutVersion)
	{
		return {
			name: packageNameWithoutVersion,
			version: packageData.version,
		}
	}
}
