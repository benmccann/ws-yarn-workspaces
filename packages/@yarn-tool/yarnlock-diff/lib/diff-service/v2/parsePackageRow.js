"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parsePackageRow = void 0;
const parsers_1 = require("@yarnpkg/parsers");
function parsePackageRow(packageName, packageData) {
    var _a, _b;
    let ret = parsers_1.parseResolution(packageData.resolution);
    let packageNameWithoutVersion = (_a = ret === null || ret === void 0 ? void 0 : ret.descriptor) === null || _a === void 0 ? void 0 : _a.fullName;
    let version = (_b = ret === null || ret === void 0 ? void 0 : ret.descriptor) === null || _b === void 0 ? void 0 : _b.description;
    if (packageNameWithoutVersion) {
        version = version.replace(/^(npm):/, '');
        if (!version.length) {
            version = ret.descriptor.description;
        }
        return {
            name: packageNameWithoutVersion,
            version,
        };
    }
}
exports.parsePackageRow = parsePackageRow;
//# sourceMappingURL=parsePackageRow.js.map