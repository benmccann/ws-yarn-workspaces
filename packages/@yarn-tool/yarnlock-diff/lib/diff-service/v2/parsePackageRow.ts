import { parseYarnLockRowV2 as parsePackageRow } from '@yarn-tool/yarnlock-util/lib/v2/parseYarnLockRowV2';

export { parsePackageRow }

export default parsePackageRow
