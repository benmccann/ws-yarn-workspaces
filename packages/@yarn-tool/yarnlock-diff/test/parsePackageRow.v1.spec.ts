import { parsePackageRow } from '../lib/diff-service/v1/parsePackageRow';

test(`@babel/parser@next`, () =>
{

	let actual = parsePackageRow('@babel/parser@next', {
		version: '1.0.0',
	} as any);

	expect(actual).toHaveProperty('name', '@babel/parser');
	expect(actual).toMatchSnapshot();

});

test(`@babel/parser@*`, () =>
{

	let actual = parsePackageRow('@babel/parser@*', {
		version: '1.0.0',
	} as any);

	expect(actual).toHaveProperty('name', '@babel/parser');
	expect(actual).toMatchSnapshot();

});
