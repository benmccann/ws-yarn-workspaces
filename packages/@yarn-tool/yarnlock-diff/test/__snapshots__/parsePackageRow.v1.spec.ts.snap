// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@babel/parser@* 1`] = `
{
  "name": "@babel/parser",
  "raw": "@babel/parser@*",
  "semver": "*",
  "type": "range",
  "version": "1.0.0",
}
`;

exports[`@babel/parser@next 1`] = `
{
  "name": "@babel/parser",
  "raw": "@babel/parser@next",
  "semver": "next",
  "type": "tag",
  "version": "1.0.0",
}
`;
